src/pyhf/parameters/paramview.py

```diff
@@ -1,114 +1,107 @@
+    target_slices = []
+    start = 0
+    for sl in slices:
+        stop = start + (sl.stop - sl.start)
+        target_slices.append(slice(start, stop))
+        start = stop
+
+    db = default_backend
+    ranges = []
+    for sl in target_slices:
+        ranges.append(db.astensor(range(sl.start, sl.stop)))
+    if not ranges:
+        return None
+    return _TensorViewer(ranges, names=names, batch_size=batch_size)
+
+
+def extract_index_access(
+    baseviewer, subviewer, indices,
+):
+    tensorlib, _ = get_backend()
+
+    index_selection = []
+    stitched = None
+    indices_concatenated = None
+    if subviewer:
+        index_selection = baseviewer.split(indices, selection=subviewer.names)
+        stitched = subviewer.stitch(index_selection)
+
+        # the transpose is here so that modifier code doesn't have to do it
+        indices_concatenated = tensorlib.astensor(
+            tensorlib.einsum('ij->ji', stitched)
+            if len(tensorlib.shape(stitched)) > 1
+            else stitched,
+            dtype='int',
+        )
+    return index_selection, stitched, indices_concatenated
 
 
 class ParamViewer(object):
     """
     Helper class to extract parameter data from possibly batched input
     """
 
-    def __init__(self, tensor_shape, par_map, name):
-        self.tensor_shape = tensor_shape
-        self.batch_shape = tensor_shape[:-1]
-        # for more general batch shapes we need to revisit
-        assert len(self.batch_shape) <= 1
-        self._par_map = par_map
-        self._index_selection = index_helper(
-            name, tensor_shape, self.batch_shape, self._par_map
-        )
+    def __init__(self, shape, par_map, selection):
 
-        if self._index_selection:
-            cat = default_backend.astensor(
-                default_backend.concatenate(self._index_selection, axis=-1), dtype='int'
-            )
-            # index_selection is
-            #   batched:   list of (batch_dim, slice size) tensors
-            #   unbatched: list of (slice size,) tensors
-            # concatenated is
-            #   batched: (batch_dim, sum of slice sizes)
-            #   unbatched: (sum of slice sizes, )
-            # indices_concatenated  is
-            #   batched: (sum of slice size, batch dim)
-            #   unbatched: (sum of slice size, )
-            if self.batch_shape:
-                self._indices_concatenated = default_backend.einsum('ij->ji', cat)
-            else:
-                self._indices_concatenated = cat
-
-        else:
-            self._indices_concatenated = None
-
-        last = 0
-        sl = []
-        for s in [
-            self._par_map[x]['slice'].stop - self._par_map[x]['slice'].start
-            for x in (name if isinstance(name, list) else [name])
-        ]:
-            sl.append(slice(last, last + s))
-            last += s
-        self._slices = sl
+        batch = shape[0] if len(shape) > 1 else None
```

Comment:
call this `batch_size`. Is the `shape` passed in here just the parameters and batched parameters are more than 1D? That's not obvious. What if we end up having N-dimensional parameters, and batching is N+1-dimensional?

Reply:
yes & done